src/test/java/org/folio/auth/authtokenmodule/TokenTest.java

```diff
@@ -0,0 +1,73 @@
+package org.folio.auth.authtokenmodule;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import java.text.ParseException;
+import com.nimbusds.jose.JOSEException;
+
+import org.folio.auth.authtokenmodule.tokens.AccessToken;
+import org.folio.auth.authtokenmodule.tokens.RefreshToken;
+import org.folio.auth.authtokenmodule.tokens.Token;
+import org.folio.auth.authtokenmodule.tokens.TokenValidationException;
+import org.junit.Before;
+import org.junit.Test;
+import io.vertx.core.Future;
+
+public class TokenTest {
+  private static String userUUID = "007d31d2-1441-4291-9bb8-d6e2c20e399a";
+  private static String passPhrase = "CorrectBatteryHorseStaple";
+  private static String tenant = "test-abc";
+  private static String username = "test-username";
+
+  @Before
+  public void setSigningKey() {
+    System.setProperty("jwt.signing.key", passPhrase);
+  }
+
+  @Test
+  public void accessTokenIsValidTest() throws JOSEException, ParseException, TokenValidationException {
+    var at = new AccessToken(tenant, username, userUUID);
+    String key = System.getProperty("jwt.signing.key");
+    var tokenCreator = new TokenCreator(key);
+    var encoded = at.encodeAsJWT(tokenCreator);
+
+    Future<Token> result = Token.validate(encoded, tokenCreator, null);
+
+    assertTrue(result.succeeded());
+    result.onSuccess(token -> {
+      assertTrue(token instanceof AccessToken);
+    });
+  }
+
+  @Test
+  public void accessTokenIsInvalidTest() throws JOSEException, ParseException, TokenValidationException {
+    String tokenMissingTenantClaim =
+      "{\"iat\":1637696002,\"sub\":\"test-username\",\"user_id\":\"007d31d2-1441-4291-9bb8-d6e2c20e399a\",\"type\":\"access\"}";
+    String key = System.getProperty("jwt.signing.key");
+    var tokenCreator = new TokenCreator(key);
+    String source = tokenCreator.createJWTToken(tokenMissingTenantClaim);
+
+    Future<Token> result = Token.validate(source, tokenCreator, null);
+
+    assert(result.failed());
+    result.onFailure(e -> {
```

Comment:
result.onFailure is not called until the future is completed. Therefore the unit test might end before it is called and the containing tests are not executed.
In this case the future is complete, but the construction should be avoided because code changes might change that in future.

    Future<Token> future = Token.validate(source, tokenCreator, null);

    assertThat(future.failed(), is(true));
    assertThat(future.cause(), is(instanceOf(TokenValidationException.class)));
    assertThat(((TokenValidationException) future.cause()).getHttpResponseCode(), is(500));

Or

    Token.validate(source, tokenCreator, null)
    .onComplete(testContext.asyncAssertFailure(e -> {
      assertThat(e, is(instanceOf(TokenValidationException.class)));
      assertThat(((TokenValidationException) e).getHttpResponseCode(), is(500));
    }));

For me assertThat instead of assertTrue is more readable, and gives better error messages if the assertion fails.

Reply:
It appears all validate implementation has no async stuff. But if they have (async stuff), only the latter will work.

Reply:
Julian thanks for letting me know about the hamcrest `CoreMatchers`. I do agree that it makes the intent of the tests easier to scan and made a change based on that. In this case the access token will never have async operations in its validation so I think it is better to not use `asyncAssertFailure` here, although I'm glad to know about that and will put it in my toolbox.